src/coreclr/vm/peassembly.cpp

```diff
@@ -359,19 +385,29 @@ void PEAssembly::ConvertMDInternalToReadWrite()
     // Swap the pointers in a thread safe manner.  If the contents of *ppImport
     //  equals pOld then no other thread got here first, and the old contents are
     //  replaced with pNew.  The old contents are returned.
-    if (InterlockedCompareExchangeT(&m_pMDImport, pNew, pOld) == pOld)
-    {
-        //if the debugger queries, it will now see that we have RW metadata
-        m_MDImportIsRW_Debugger_Use_Only = TRUE;
 
-        // Swapped -- get the metadata to hang onto the old Internal import.
-        HRESULT hr=m_pMDImport->SetUserContextData(pOld);
-        _ASSERTE(SUCCEEDED(hr)||!"Leaking old MDImport");
-        IfFailThrow(hr);
-    }
-    else
-    {   // Some other thread finished first.  Just free the results of this conversion.
+    if (InterlockedCompareExchangeT(&m_pConvertedMDImport, pNew, NULL) != NULL)
+    {
+        // If we lost the race, free our copy of the RW MD. However, we might still
+        // be the only requesting swapping and our caller would expect m_pMDImport to
+        // be RW. Try to set it to the converted import.
         pNew->Release();
+        pNew = m_pConvertedMDImport;
+    }
+
+    if (swapForRWMDImport)
+    {
+        if (InterlockedCompareExchangeT(&m_pMDImport, pNew, pOld) == pOld)
+        {
+            //if the debugger queries, it will now see that we have RW metadata
+            m_MDImportIsRW_Debugger_Use_Only = TRUE;
```

Comment:
Do we need to set `m_MDImportIsRW_Debugger_Use_Only` in the other two places that call SetUserContextData?

Reply:
There's only one other I could find, and yes. Thanks for the catch.